lib/protocol/Auth.js

```diff
@@ -44,6 +59,29 @@ Auth.token = function(password, scramble) {
   return xor(stage3, stage1);
 };
 
+Auth.sha2Token = function(password, scramble) {
+  if (!password) {
+    return Buffer.alloc(0);
+  }
+
+  // password must be in binary format, not utf8
+  var stage1 = sha256((Buffer.from(password, 'utf8')).toString('binary'));
+  var stage2 = sha256(stage1);
+  var stage3 = sha256(stage2 + scramble.toString('binary'));
+  return xor(stage1, stage3);
+};
+
+Auth.encrypt = function(password, scramble, key) {
+  if (typeof Crypto.publicEncrypt !== 'function') {
+    var err  = new Error('The Node.js version does not support public key encryption');
+    err.code = 'PUB_KEY_ENCRYPTION_NOT_AVAILABLE';
+    throw err;
+  }
+
+  var stage1 = xor((Buffer.from(password + '\0', 'utf8')).toString('binary'), scramble.toString('binary'));
```

Comment:
this needs to be changed to rolling xor, otherwise it'll fail for password over 20 characters long

See discussion at https://github.com/sidorares/node-mysql2/issues/1044

Reply:
Yea, I've been working to fix this PR up over the past week and I did notice that long passwords didn't work on the new auth. There are a bunch of other little minor issues I've been finding as well. I didn't intend to list them out and was just intending to push up all the changes here then merge, but let me know if you think I should list them all out in addition to pushing up the fixes.

Reply:
Note that mysql2 is using slightly different api for this. No config required, and if you need to use non-defaults you pass custom configured plugin as `authPlugins: { caching_sha2_password: XXXX }`. Should not be a big problems for user migrating both ways mysql<->mysql2